### PR TITLE
Fixing Event Page When section formatting

### PR DESCRIPTION
**Pull request by moanu404**
# Description
This PR creates a new date formatter function that addresses a defect currently leading to showing of the date in the "When" section inconsistently and it formats the date/time and time zone to be consistent with the AC. I tried to keep the formatting as consistent as possible with the format on the event listing page.
Now also using date-fns, date-fns-tz library
## Ticket
Relates to #[19696](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/19696)

## Developer Task

```
- [ ] PR submitted against the `main` branch of `next-build`.
- [ ] Link to the issue that this PR addresses (if applicable).
- [ ] Define all changes in your PR and note any changes that could potentially be breaking changes.
- [ ] PR includes steps to test your changes and links to these changes in the Tugboat preview (if applicable).
- [ ] Provided before and after screenshots of your changes (if applicable).
- [ ] Alerted the #accelerated-publishing Slack channel to request a PR review.
- [ ] You understand that once approved, you are responsible for merging your changes into `main`. (Note that changes to `main` will move automatically into production.)
```

## Testing Steps
- [ ] Run all unit tests `yarn test`
- [ ] Load up https://www.va.gov/outreach-and-events/events 
- [ ] Click into the event and observe the "When:" area and see the incorrect formatting 
- [ ] EG https://www.va.gov/dayton-health-care/events/76691/ 
- [ ] Load up your local server and pull down this branch
- [ ] check http://localhost:3999/outreach-and-events/events/ (takes forever) OR http://localhost:3999/dayton-health-care/events/
- [ ] do the same parity check and ensure the formatting is consistent and so is the time zone is in your local timezone

## QA steps
What needs to be checked to prove this works?  
What needs to be checked to prove it didn't break any related things?  
What variations of circumstances (users, actions, values) need to be checked?

## Screenshots
Before: (PROD)
image

After: (Local)
image

## Is this PR blocked by another PR?
- Add the `DO NOT MERGE` label
- Add links to additional PRs here:

# Reviewer

### Reviewing a PR

This section lists items that need to be checked or updated when making changes to this repository. 

## Standard Checks

```
- [ ] Code Quality: Readabilty, Naming Conventions, Consistency, Reusability
- [ ] Test Coverage: 80% coverage
- [ ] Functionality: Change functions as expected with no additional bugs
- [ ] Performance: Code does not introduce performance issues
- [ ] Documentation: Changes are documented in their respective README.md files
- [ ] Security: Packages have been approved in the TRM
```

## Merging an Approved Layout

When merging a layout, you must ensure that the content type has been turned on for `next-build` inside the `CMS`. This CMS flag must be turned on for editors to preview their work using the next build preview server.

Resource types (layouts) that have not been approved by design should NOT be pushed to production. Ensure that [slug.tsx](../src/pages/[[...slug]].tsx) does not include your resource type if it is not approved.

The status of layouts should be kept up to date inside [templates.md](./templates.md). This includes QA progress, development progress, etc. A link should be provided for where testing can occur.

## Merging a Non-Approved Layout

Your new resource type should not be included inside [slug.tsx](../src/pages/[[...slug]].tsx). Items added here will go into production once merged into the `main` branch. It is imperative that we do not push anything live that has not been approved.

Ensure that this layout has been added to the [templates.md](./templates.md) file with the current status of the work.

### CODE REVIEW

File: src/lib/utils/date.test.ts [hunk 1]
```diff
@@ -342,3 +343,48 @@ test('getDateParts should return null for falsy values', () => {
 test('parseDate should return null for inputs that do not match any date format', () => {
   expect(parseDate('')).toBe(null)
 })
+
+describe('formatEventDateTime', () => {
+  beforeEach(() => {
+    // Check if we're in UTC (CI environment)
```
Comment (moanu404) [inline src/lib/utils/date.test.ts:349]: this makes me sad but the tests kept failing in CI cuz UTC

File: src/lib/utils/date.ts [hunk 2]
```diff
@@ -257,6 +257,36 @@ export const deriveFormattedTimestamp = (datetime) => {
...
 
+import { format } from 'date-fns'
+
+interface DateTimeRange {
+  value: number
+  endValue: number
+  timezone?: string
+}
+
+export const formatEventDateTime = (mostRecentDate: DateTimeRange | null) => {
```
Comment (moanu404) [inline src/lib/utils/date.ts:268]: Tried to make this more or less adherent to vets-website except not using moment

**Comment by moanu404** [comment]
@laflannery while tugboat is being flakey on me I had a quick question about formatting
https://github.com/department-of-veterans-affairs/next-build/pull/910 here we can see that 12 pm should be noon and 12 am midnight etc, should we account for this here?

**Comment by laflannery** [comment]
@moanu404 The formatting looks great! A couple comments:

1. Regarding your noon/midnight question, we weren't doing this on Prod prior to the migration so it's ok if we use 12:00 am/pm here as well. I actually think it makes a bit more sense to use the numeric times here.
2. On your stories PR, you changed [this common file](https://github.com/department-of-veterans-affairs/next-build/blob/main/src/templates/common/socialLinks/index.tsx) and it looks like it updated the spacing on the social links in the Events detail pages. I didn't catch it because I didn't realize it was using a shared component. Can you take a look? This is a screenshot from a previous ticket to show what it should look like (ignore the red lines, those were for the purpose of the original ticket I took this from):
image

**Comment by laflannery** [comment]
@moanu404 I re-reviewed the recurring event times and those look good now, thanks!

**Comment by laflannery** [comment]
After discussion with @moanu404 we decided to go ahead and merge this and handle the spacing issue I noted above in [a separate ticket](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/20874)

File: package.json [hunk 3]
```diff
@@ -56,6 +56,7 @@
     "chrome-launcher": "^1.1.2",
     "clsx": "^2.1.0",
     "commander": "^12.0.0",
+    "date-fns": "^4.1.0",
```
Comment (cvalarida) [inline package.json:59]: https://blog.date-fns.org/v40-with-time-zone-support/ 🎉 

This is weirdly exciting. 😂 🤓

File: src/lib/utils/date.test.ts (hunk 1 above)
Comment (cvalarida) [reply src/lib/utils/date.test.ts:349]: 🤔 I don't think I understand why this is necessary. If we're in UTC (CI), we say the mock time zone is EST, but if it isn't (including for actual EST), we say the mock time zone is PST. Why not stick with one or the other regardless of what the system time zone is?

How was this failing?

File: src/lib/utils/date.test.ts [hunk 4]
```diff
@@ -342,3 +343,48 @@ test('getDateParts should return null for falsy values', () => {
...
+  it('formats date and time with timezone', () => {
+    const mostRecentDate = {
+      value: 1683216000,
+      endValue: 1683223200,
+    }
+
+    const result = formatEventDateTime(mostRecentDate)
+    const [datePart, timePart] = result.split(' – ')
+    expect(datePart).toMatch(/Thu\. May 4, 2023, \d{1,2}:\d{2} [ap]\.m\./)
+    expect(timePart).toMatch(/\d{1,2}:\d{2} [ap]\.m\. [A-Z]T/)
```
Comment (cvalarida) [inline src/lib/utils/date.test.ts:385]: I would expect this to expect a specific value, since we're providing a specific value. Like, either it's `9:00 a.m. - 11:00 a.m. PT` or `noon - 3:00 p.m. ET`.

Can we add tests to check that the time zone is being translated correctly as well as the a.m. / p.m. / noon / midnight? This test checks that the format _looks_ correct, not that the times actually _are_ correct.

File: src/lib/utils/date.ts [hunk 5]
```diff
@@ -257,6 +257,36 @@ export const deriveFormattedTimestamp = (datetime) => {
   return reformatTime(formattedTime)
 }
 
+import { format } from 'date-fns'
+
+interface DateTimeRange {
+  value: number
+  endValue: number
+  timezone?: string
+}
```
Comment (cvalarida) [inline src/lib/utils/date.ts:266]: Is this the same shape as in `vets-website`? If not, I recommend renaming `value` and `endValue` to `startTime` and `endTime`. In either case, I recommend putting a JSDoc comment on the properties to describe them.
```suggestion
/**
 * Bonus points for describing this object, but the name is pretty descriptive by itself. 🤷‍♂️
 */
interface DateTimeRange {
  /**
   * Unix timestamp for the beginning of the date time range.
   */
  value: number
  /**
   * Unix timestamp for the end of the date time range.
   */
  endValue: number
  /**
   * ISO timezone string (e.g. 'America/Los_Angeles')
   */
  timezone?: string
}
```

It feels pretty verbose, but we get to leverage the power of our language server hover docs that way. 🤓

File: src/lib/utils/date.ts (hunk 2 above)
Comment (cvalarida) [reply src/lib/utils/date.ts:268]: Can you add a JSDoc comment to this as well? I'd like to see at a glance what the format is and call out that this is _specifically_ for formatting dates and times for _events_. Bonus points for linking to the [design system content style guide for dates and time](https://design.va.gov/content-style-guide/dates-and-numbers#times-and-time-zones) for reference.

**Pro tip:** Get the AI to write it for you, then correct any mistakes it may make. 🤓 

We might could extract this formatting logic later since it seems more broadly applicable than for _just_ events.

File: src/lib/utils/date.ts [hunk 6]
```diff
@@ -257,6 +257,36 @@ export const deriveFormattedTimestamp = (datetime) => {
...
+  if (!mostRecentDate) return ''
+
+  const startsAtUnix = mostRecentDate.value
+  const endsAtUnix = mostRecentDate.endValue
+  const formattedStartsAt = startsAtUnix
+    ? format(new Date(startsAtUnix * 1000), 'EEE. MMM d, yyyy, h:mm aaaa')
+    : ''
+  const formattedEndsAt = endsAtUnix
+    ? format(new Date(endsAtUnix * 1000), 'h:mm aaaa')
+    : ''
```
Comment (cvalarida) [inline src/lib/utils/date.ts:278]: Alright, so this gets a little funky.

1. If I pass in `formatEventDateTime({ value: 0, endValue: 0 })`, it looks like the output is going to be ` –  PT`; is that right?
2. Neither of these account for the `noon` or `midnight` edge cases as described by the style guide

File: src/lib/utils/date.ts [hunk 7]
```diff
@@ -257,6 +257,36 @@ export const deriveFormattedTimestamp = (datetime) => {
...
+  const endsAtTimezone =
+    new Intl.DateTimeFormat('en-US', {
+      timeZoneName: 'short',
+    })
+      .formatToParts(new Date())
+      .find((part) => part.type === 'timeZoneName')
+      ?.value?.replace(/S|D/i, '') || ''
```
Comment (cvalarida) [inline src/lib/utils/date.ts:285]: > When referencing international time zones, use the time zone name for that country or the UTC (coordinated universal time) offset. (Example: Central European Time; Korea Standard Time; UTC +2)

A couple things here:
1. `timeZoneName: 'short'` is apparently _only_ for US timezones (maybe only continental US; I don't know)
2. If we just blindly remove the first `S` or `D` like this `.replace()` does _and_ we use a time zone outside of the US, we may end up with a formatted timezone that looks like `Korea tandard Time` 😅

File: src/lib/utils/date.ts [hunk 8]
```diff
@@ -257,6 +257,36 @@ export const deriveFormattedTimestamp = (datetime) => {
...
+
+  return `${formattedStartsAt} – ${formattedEndsAt} ${endsAtTimezone}`
```
Comment (cvalarida) [inline src/lib/utils/date.ts:287]: > - In body copy, write out ranges using sentence construction: We’re open 9:00 a.m. to 6:00 p.m. ET, Monday through Friday.
> - Where space is limited, indicate ranges with the - hyphen without spaces on either side. Don’t combine the hyphen with from/between sentence construction.

Not sure if this function will be used in these cases, but food for thought: What would this look like if we added an `options?` parameter that we could use to configure the output based on where we're using it?

I think until we run into a situation where we for sure want to use `formatEventDateTime`, we shouldn't worry about it, but food for thought. :homer-disappear:

**Review by cvalarida — COMMENTED** [review]
Looks like a solid start! I think there are some edge cases we need to consider, and the testing needs to be a little more robust.

File: src/lib/utils/date.ts (hunk 6 above)
Comment (moanu404) [reply src/lib/utils/date.ts:278]: 1. Good catch, I can definitely add some error handling around that, however would that be a valid input given from what I am looking up a timestamp of 0 means January 1, 1970 00:00:00 UTC and given in the VA events system we only look for events in the future (everything else shows up as an invalid date atm in prod) the scenario seems unlikely. Does not hurt to add it though either ways. 

2. So I confirmed with Laura, and she commented here too -  that current behavior in Prod, does not have noon/midnight so it isn't necessary to implement here

File: src/lib/utils/date.ts (hunk 7 above)
Comment (moanu404) [reply src/lib/utils/date.ts:285]: Oh yes the code I based this off seemed to be consistent from the code i was referring to on `vets-website` like which appeared to be US specific  -  
https://github.com/department-of-veterans-affairs/vets-website/blob/5c0d4b46de7f87095b88d44146092c09ec583acf/src/applications/static-pages/events/components/Results/index.js#L84

This is getting more funky (when have dates not been lol) But if we are using end users timezones now, and someone is not logging in from within US from the style guide i suppose we can display a UTC offset?

File: src/lib/utils/date.ts (hunk 8 above)
Comment (moanu404) [reply src/lib/utils/date.ts:287]: thats a good thought for sure if we see another use case for it, until then, it lives in the utils class if we ever need it again

File: src/lib/utils/date.test.ts (hunk 1 above)
Comment (moanu404) [reply src/lib/utils/date.test.ts:349]: I was trying this a bit ago and I was checking against a hardcoded string and it kept failing against it, I can spend some more time to brainstorm alternate ways to better ways to test this but I was hitting a wall against the UTC timezone results going against my expects and failing the CI

File: src/lib/utils/date.ts (hunk 6 above)
Comment (cvalarida) [reply src/lib/utils/date.ts:290]: 1. :party-wizard:
2. :design-system-twitch-intensifies:

File: package.json (hunk 3 above)
Comment (moanu404) [reply package.json:59]: I used `date-fns-tz` for refactoring the edge cases since I found it easier to do our funky formatting and it had more wider usage 😅

File: src/lib/utils/date.test.ts [hunk 9]
```diff
@@ -342,3 +350,90 @@ test('getDateParts should return null for falsy values', () => {
...
+    }
+    const result = formatEventDateTime(mostRecentDate)
+    expect(result).toBe('Invalid date – Invalid date ET')
+  })
+
+  it('should format event in US timezone if user is international (Asia)', () => {
+    mockUserTimeZone('Asia/Tokyo')
+    const mostRecentDate = {
+      value: 1683201600,
+      endValue: 1683208800,
```
Comment (moanu404) [inline src/lib/utils/date.test.ts:422]: AI helped a lot to get these stamps 🤖

**Comment by moanu404** [comment]
New change is such that `mostRecentDate` timezone will now be used if the system user is located internationally, and if they are in the US, then the event times will be shown in their timezone

**Comment by laflannery** [comment]
Looks good to me, I tested a couple international timezones and saw CT and PT for the events as expected

**Review by cvalarida — APPROVED** [review]
I apparently never actually submitted this. 🫠 

I've got a few questions / comments, but nothing merge-blocking, I think.

File: package.json (hunk 3 above)
Comment (moanu404) [reply package.json:59]: Oops posted this in the wrong spot since I have so many PR's open lol
So I tried that initially and it did not work (I cannot pull up that terminal log alas) but it seemed to require me to have `date-fns` in addition for it to pass, which was weird. I'll try to dig it up some more and see what exact error it was showing me. The vets-website counterpart is now using `date-fns-tz`  and I saw more examples with formatting with it is what I meant.

File: src/lib/utils/date.test.ts [hunk 10]
```diff
@@ -1,3 +1,10 @@
+/* eslint-disable import/no-extraneous-dependencies */
+import { jest } from '@jest/globals'
+
+jest.mock('date-fns-tz', () => ({
+  formatInTimeZone: jest.fn(),
+}))
```
[the remark this answers is not in the document]
Comment (moanu404) [reply src/lib/utils/date.test.ts:6]: Closed my terminal window (argh) but the pre-commit linter was unhappy about the presence of jest in a test file, so I had to add the `/* eslint-disable import/no-extraneous-dependencies */` as far as I can recall

File: src/lib/utils/date.ts [hunk 11]
```diff
@@ -257,6 +257,96 @@ export const deriveFormattedTimestamp = (datetime) => {
...
+
+  // Format start and end times in target timezone
+  const formattedStartsAt =
+    startsAtUnix && startsAtUnix > 0
+      ? formatInTimeZone(
+          new Date(startsAtUnix * 1000),
+          targetTimeZone,
+          'EEE. MMM d, yyyy, h:mm aaaa'
+        )
+      : 'Invalid date'
```
[the remark this answers is not in the document]
Comment (laflannery) [reply src/lib/utils/date.ts:333]: FYI - Past Events cannot be reached unless:
- You know the URL exactly and can go directly there or
- You search and the event shows up in the search results
However that is only temporary, eventually past events won't be on the site or reachable at all

[the remark this answers is not in the document]
Comment (laflannery) [reply src/lib/utils/date.ts:333]: Ya - if it's for overall monitoring I think that makes sense but if it's just for the past events Invalid date I don't think it makes sense, only because they are going away. But yes, I think a ticket makes sense and then we can discuss and refine it, thanks!
